torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -912,3 +923,182 @@ def save_checkpoint(
+            if re.match(checkpoint_dir_pattern, name)
+            and os.path.isfile(
+                os.path.join(self._output_dir, name, self._metadata_file)
+            )
+        ]
+
+        if checkpoint_paths:
+            latest_checkpoint_dir = sorted(checkpoint_paths)[-1]
+            return os.path.join(self._output_dir, latest_checkpoint_dir)
+        return None
+
+    def load_checkpoint(self, state_dict: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Load a Distributed checkpoint
+        """
+        checkpoint_path = self._get_latest_intermediate_checkpoint_path()
+
+        if checkpoint_path is None:
+            raise ValueError(
+                "No intermediate checkpoint found in the specified directory. Please ensure that a checkpoint exists."
+            )
+
+        if self._is_rank_zero:
+            logger.info(f"Loading checkpoint from {checkpoint_path}")
+
+        load(
+            state_dict,
+            storage_reader=FileSystemReader(checkpoint_path),
+        )
+
+        return state_dict
+
+    def save_checkpoint(
+        self,
+        state_dict: Dict[str, Any],
+        epoch: int,
+        save_async: bool = False,
+    ) -> None:
+        """
+        Save a distributed checkpoint to disk.
+        If ``save_async`` is True, the save happens asynchronously unblocking the GPUs faster.
+
+        Args:
+            state_dict (Dict[str, Any]): Checkpoint state dict to be written out to file
+            epoch (int): Epoch number. Used to create the checkpoint file name
+            save_async (bool): If True, save the checkpoint asynchronously
+        """
+
+        if self._is_rank_zero:
+            logger.info(
+                f"DistributedCheckpointer is saving a checkpoint for the epoch {epoch}"
+            )
+
+        checkpoint_path = Path.joinpath(
+            self._output_dir, f"{self._checkpoint_dir_prefix}_{epoch}"
+        )
+
+        if self._checkpoint_future is not None:
+            # Previous checkpoint needs to finish before saving the next one.
+            wait_start = time.perf_counter()
+
+            if self._is_rank_zero:
+                logger.info(
+                    "Previous checkpoint has not finished. Checkpointing frequency is too high. Waiting..."
+                )
+
+            self._checkpoint_future.result()
+
+            if self._is_rank_zero:
+                logger.info(
+                    f"Waited {time.perf_counter() - wait_start:.2f} seconds for previous checkpoint to finish"
+                )
+
+        cp_start = time.perf_counter()
+
+        if save_async:
+
+            def callback(
+                f: Future,
+            ) -> None:
+                if f.exception() is None:
+                    if self._is_rank_zero:
+                        logger.info(
+                            f"Checkpoint is saved asynchronously to {checkpoint_path}"
+                        )
+
+                    for index in range(epoch):
```

Comment:
also nit but do you really need a for loop here if this runs every epoch? (i.e. isn't it just sufficient to clean up epoch n-1 checkpoint after each epoch n save)

Reply:
True. We can do that. I added this code instead just for edge cases when some prior checkpoint folder still did not get cleaned up due to user cancelling the job or the job errored out while the clean up was happening asynchronously.